### PR TITLE
Bug in whitelisting logic

### PR DESCRIPTION
**Pull request by andrewhancox**
There is a bug that means that if you have a whitelist set then Its that don't match the whitelist get rejected even if they match the value on the key. Have added test to check for this scenario.

### CODE REVIEW

**Comment by dmitriim** [comment]
Hi @andrewhancox ! 
Thanks of your contribution! Can you please provide some more information on this bug? Maybe some examples. It seems like I can't easily get what the bug is about and what the fix does.

**Comment by andrewhancox** [comment]
The problem was that if you supplied any value for the whitelist it only compared the incoming IP against that, whereas it should also still compare it against the IP that the key has been locked to.

File: classes/core_userkey_manager.php [hunk 1]
```diff
@@ -122,38 +122,45 @@ public function validate_key($keyvalue) {
...
-            } else if (!address_in_subnet($remoteaddr, $key->iprestriction)) {
-                print_error('ipmismatch', 'error', '', null, "Remote address: $remoteaddr\nKey IP: $key->iprestriction");
-            }
-        }
-
         if (!$user = $DB->get_record('user', array('id' => $key->userid))) {
             print_error('invaliduserid');
         }
 
+        $this->validate_ip_address($key);
```
Comment (dmitriim) [inline classes/core_userkey_manager.php]: Why did you move this validation down after getting user from DB?

Comment (andrewhancox) [reply classes/core_userkey_manager.php]: That wasn't intentional, I have changed the order back.
